src/Application/Commands/RedistributeMatchFunds.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MatchBot\Application\Commands;
+
+use MatchBot\Domain\CampaignFundingRepository;
+use MatchBot\Domain\DonationRepository;
+use MatchBot\Domain\FundingWithdrawal;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Redistribute match funding allocations where possible, from lower to higher priority match fund pots.
+ */
+class RedistributeMatchFunds extends LockingCommand
+{
+    protected static $defaultName = 'matchbot:redistribute-match-funds';
+
+    public function __construct(
+        private CampaignFundingRepository $campaignFundingRepository,
+        private DonationRepository $donationRepository
+    ) {
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this->setDescription('Moves match funding allocations from lower to higher priority funds where possible');
+    }
+
+    protected function doExecute(InputInterface $input, OutputInterface $output): int
+    {
+        // TODO Change the fixed lookback to 2 days, or parameter-ise it, once CC23 is tidied in Prod.
+        $donationsToCheck = $this->donationRepository->findWithMatchingWhichCouldBeReplacedWithHigherPriorityAllocation(
+            new \DateTimeImmutable('-3 weeks')
+        );
+
+        $donationsAmended = 0;
+        foreach ($donationsToCheck as $donation) {
+            $highestAllocationOrderUsedForDonation = 0;
+            foreach ($donation->getFundingWithdrawals() as $withdrawal) {
+                $funding = $withdrawal->getCampaignFunding();
+                if (!$funding) {
+                    // Not entirely sure why this is nullable at all; we can't work on withdrawals without a funding.
```

Comment:
There are no cases in the prod database where it is set to null: `SELECT FundingWithdrawal.campaignFunding_id FROM FundingWithdrawal WHERE campaignFunding_id is null; ` returns empty. 

So can we make it non-nullable instead of adding code to deal with the case of it being null? There's only one call to `$withdrawal = new FundingWithdrawal();` in prod code, and it has a call to `$withdrawal->setCampaignFunding($funding);` just after so it looks like we could easily make it a required constructor param.

Reply:
Although probably worth looking into the reason it's nullable - it's this commit: https://github.com/thebiggive/matchbot/commit/a9d71f4eed239c84dd2082eb2ddb410817fc7050

Reply:
I checked that commit too but I can't tell what we had a problem with before. I think our testing is good enough now that if things are green we can probably assume it's no longer an issue. Changed in 2a2b0988a8daa37b4564257c5f6acb14f7c0598e